message_ix_models/model/workflow.py

```diff
@@ -0,0 +1,99 @@
+"""Common steps for workflows."""
+
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Any, Dict, Optional
+
+from message_ix import Scenario
+
+from message_ix_models.util.config import ConfigHelper
+
+if TYPE_CHECKING:
+    from message_ix_models import Context
+
+
+@dataclass
+class Config(ConfigHelper):
+    """Common configuration for model workflows across projects.
+
+    Currently, the three settings are understood by :func:`solve`, which is used in
+    ENGAGE, NAVIGATE, and possibly other workflows.
+    """
+
+    #: Information on an optional, other scenario from which to copy demand data in
+    #: :func:`solve` using :func:`transfer_demands`. Default: empty, do nothing.
+    demand_scenario: Dict = field(default_factory=dict)
+
+    #: :obj:`True` to call :func:`.reserve_margin.res_marg.main` in :func:`solve`.
+    reserve_margin: bool = True
+
+    #: Keyword arguments for :meth:`.message_ix.Scenario.solve` via :func:`solve`.
+    #:
+    #: To replicate the behaviour of the `macro_params` argument to
+    #: :meth:`.engage.ScenarioRunner.run`, which in turn sets the `convergence_issues`
+    #: argument to :meth:`.engage.ScenarioRunner.solve`, set max_adjustment to 0.1.
+    solve: Dict[str, Any] = field(
+        default_factory=lambda: dict(model="MESSAGE-MACRO", max_adjustment=0.2)
+    )
+
+
+def solve(
+    context: "Context",
+    scenario: Scenario,
+    *,
+    config: Optional[Config] = None,
+    set_as_default: bool = False,
+):
+    """Common model solve step for ENGAGE, NAVIGATE, and other workflows.
+
+    The steps respond settings from an optional instance of :class:`Config` passed as a
+    keyword argument.
+
+    *Before* `scenario` is solved:
+
+    1. If :attr:`Config.reserve_margin` is :obj:`True`, :func:`.res_marg.main` is
+       called.
+    2. If :attr:`Config.demand_scenario` is non-empty and ``config.solve["model"]`` is
+       ``MESSAGE-MACRO``, then :func:`.transfer_demands` is called to transfer data from
+       the indicated scenario to `scenario`. This scenario **must** exist on the same
+       :class:`.Platform` as `scenario`; the default version is loaded.
+
+    Then:
+
+    3. `scenario` is solved passing :attr:`Config.solve` as keyword arguments to
+       :meth:`.message_ix.Scenario.solve`. The keyword argument `var_list` has ``I``,
+       ``C``, and (if ``config.solve["model"]`` is ``MESSAGE-MACRO``) ``GDP`` appended.
+
+    *After* `scenario` is solved without exception:
+
+    4. :meth:`.set_as_default` is called if the keyword argument `set_as_default` is
+       :obj:`True`.
+    """
+
+    from message_data.scenario_generation.reserve_margin import res_marg
```

Comment:
Please check that the code doesn't import from message_data anymore, if at all possible.

Reply:
This is a difficult case: we will likely not want to port `message_data.scenario_generation` as-is since (at least judging by its name) shadows existing modules like `message_ix_models.model.bare`, `message_ix_models.model.structure` etc.

I guess this particular function might go in `.tools` or `.util`, but it needs clean-up, added tests, etc. —all of which I decided were out-of-scope for this PR.

I've opened #211 to address.

Reply:
Sounds fine for me. Let's hope no one tries to use this particular function until #211 is closed, which should ideally happen sooner rather than later.